File: swift/Sources/MLXAudioModels/CLAP/FeatureExtractor.swift
// FeatureExtractor.swift
// Audio feature extraction for CLAP.
//
// Implements audio preprocessing compatible with HuggingFace's ClapFeatureExtractor,
// converting raw waveforms to mel spectrograms suitable for the CLAP audio encoder.

import Foundation
@preconcurrency import MLX
import MLXAudioPrimitives

// MARK: - CLAP Feature Extractor

/// Audio feature extractor for CLAP models.
///
/// Converts audio waveforms to log-mel spectrograms with the specific
/// preprocessing required by CLAP:
/// - 48kHz sample rate
/// - 64 mel bins
/// - 10ms hop length
/// - Up to 10 seconds of audio (with fusion for longer)
public struct CLAPFeatureExtractor: Sendable {
    /// Target sample rate.
    public let sampleRate: Int

    /// FFT window size.
    public let nFFT: Int

    /// Hop length for STFT.
    public let hopLength: Int

    /// Number of mel bins.
    public let nMels: Int

    /// Maximum audio length in seconds.
    public let maxLengthS: Float

    /// Maximum number of samples.
    public var maxSamples: Int {
        Int(maxLengthS * Float(sampleRate))
    }

    /// Expected number of frames for max length audio.
    public var maxFrames: Int {
        maxSamples / hopLength + 1
    }

    /// Creates a feature extractor with the given configuration.
    public init(
        sampleRate: Int = 48000,
        nFFT: Int = 1024,
        hopLength: Int = 480,
        nMels: Int = 64,
        maxLengthS: Float = 10.0
    ) {
        self.sampleRate = sampleRate
        self.nFFT = nFFT
        self.hopLength = hopLength
        self.nMels = nMels
        self.maxLengthS = maxLengthS
    }

    /// Create feature extractor from CLAP audio config.
    public static func fromConfig(_ config: CLAPAudioConfig) -> CLAPFeatureExtractor {
        CLAPFeatureExtractor(
            sampleRate: config.sampleRate,
            nFFT: config.nFFT,
            hopLength: config.hopLength,
            nMels: config.nMels,
            maxLengthS: config.maxLengthS
        )
    }

    /// Process audio waveform into CLAP-compatible features.
    ///
    /// - Parameters:
    ///   - audio: Audio waveform [T] or [B, T]
    ///   - inputSampleRate: Sample rate of input audio (resamples if different)
    /// - Returns: Tuple of (inputFeatures, isLonger)
    ///   - inputFeatures: Mel spectrogram [B, 4, F, T] for fusion mode
    ///   - isLonger: Boolean array [B] indicating which samples need fusion
    public func callAsFunction(
        _ audio: MLXArray,
        inputSampleRate: Int? = nil
    ) throws -> (inputFeatures: MLXArray, isLonger: MLXArray) {
        var processedAudio = audio

        // Handle 1D input
        if processedAudio.ndim == 1 {
            processedAudio = processedAudio.expandedDimensions(axis: 0)
        }

        let batchSize = processedAudio.shape[0]
        let originalLength = processedAudio.shape[1]

        // Resample if needed
        if let sr = inputSampleRate, sr != sampleRate {
            processedAudio = resample(processedAudio, fromRate: sr, toRate: sampleRate)
        }

        let currentLength = processedAudio.shape[1]

        // Determine which samples are longer than max
        let isLongerBool = currentLength > maxSamples
        let isLongerArray = MLXArray(
            Array(repeating: isLongerBool ? Float(1) : Float(0), count: batchSize)
        ).reshaped([batchSize, 1])

        // Process each sample
        var allFeatures: [MLXArray] = []

        for b in 0..<batchSize {
            let sample = processedAudio[b]
            let sampleLength = sample.shape[0]

            let mel: MLXArray
            let channel0: MLXArray  // Global/downsampled
            let channel1: MLXArray  // Local chunk 1
            let channel2: MLXArray  // Local chunk 2
            let channel3: MLXArray  // Local chunk 3

            if sampleLength > maxSamples {
                // Long audio: create fusion features
                let fusionResult = try computeFusionMel(sample)
                channel0 = fusionResult.global
                channel1 = fusionResult.local1
                channel2 = fusionResult.local2
                channel3 = fusionResult.local3
            } else {
                // Short audio: pad/repeat and duplicate
                let paddedMel = try computePaddedMel(sample)
                channel0 = paddedMel
                channel1 = paddedMel
                channel2 = paddedMel
                channel3 = paddedMel
            }

            // Stack channels: [4, F, T]
            let stacked = MLX.stacked([channel0, channel1, channel2, channel3], axis: 0)
            allFeatures.append(stacked.expandedDimensions(axis: 0))
        }

        // Concatenate batch: [B, 4, F, T]
        let inputFeatures = MLX.concatenated(allFeatures, axis: 0)

        return (inputFeatures, isLongerArray)
    }

    /// Compute log-mel spectrogram.
    private func computeMel(_ audio: MLXArray) throws -> MLXArray {
        // Create mel config for CLAP
        let melConfig = MelConfig(
            sampleRate: sampleRate,
            nMels: nMels,
            fMin: 50.0,
            fMax: 14000.0,
            htk: true,  // CLAP uses HTK mel scale
            norm: .slaney
        )

        // Compute mel spectrogram
        var mel = try melspectrogram(
            audio,
            nFFT: nFFT,
            hopLength: hopLength,
            melConfig: melConfig,
            power: 2.0
        )

        // Convert to dB
        mel = powerToDb(mel, ref: 1.0, amin: 1e-10, topDb: 80.0)

        return mel  // [nMels, nFrames]
    }

    /// Compute mel for short audio with padding.
    private func computePaddedMel(_ audio: MLXArray) throws -> MLXArray {
        var processedAudio = audio
        let length = processedAudio.shape[0]

        // Pad with repeat if too short
        if length < maxSamples {
            let nRepeat = (maxSamples + length - 1) / length
            var repeated = processedAudio
            for _ in 1..<nRepeat {
                repeated = MLX.concatenated([repeated, processedAudio], axis: 0)
            }
            processedAudio = repeated
        }

        // Truncate or pad to exact length
        let currentLength = processedAudio.shape[0]
        if currentLength > maxSamples {
            processedAudio = processedAudio[0..<maxSamples]
        } else if currentLength < maxSamples {
            let padding = MLX.zeros([maxSamples - currentLength])
            processedAudio = MLX.concatenated([processedAudio, padding], axis: 0)
        }

        return try computeMel(processedAudio)
    }

    /// Compute fusion mel for long audio.
    private func computeFusionMel(_ audio: MLXArray) throws -> (
        global: MLXArray,
        local1: MLXArray,
        local2: MLXArray,
        local3: MLXArray
    ) {
        // Compute full mel
        let fullMel = try computeMel(audio)  // [nMels, totalFrames]
        let totalFrames = fullMel.shape[1]
        let chunkFrames = maxFrames

        // If audio is shorter than expected, just repeat
        if chunkFrames >= totalFrames {
            return (fullMel, fullMel, fullMel, fullMel)
        }

        // Get 3 random crops from different parts
        let rangeSize = (totalFrames - chunkFrames) / 3

        // Front chunk
        let frontStart = 0
        let frontEnd = min(frontStart + chunkFrames, totalFrames)
        let local1 = fullMel[0..., frontStart..<frontEnd]

        // Middle chunk
        let middleStart = rangeSize
        let middleEnd = min(middleStart + chunkFrames, totalFrames)
        let local2 = fullMel[0..., middleStart..<middleEnd]

        // Back chunk
        let backStart = 2 * rangeSize
        let backEnd = min(backStart + chunkFrames, totalFrames)
        let local3 = fullMel[0..., backStart..<backEnd]

        // Global: downsample full mel to chunkFrames
        // Transpose to [1, 1, F, T] for interpolation
        let fullMelExpanded = fullMel.expandedDimensions(axes: [0, 1])
        let globalMel = interpolate2DSimple(
            fullMelExpanded,
            targetShape: (nMels, chunkFrames)
        )
        let global = globalMel.squeezed(axes: [0, 1])

        return (global, local1, local2, local3)
    }

    /// Simple 2D bilinear interpolation for mel downsampling.
    private func interpolate2DSimple(_ x: MLXArray, targetShape: (Int, Int)) -> MLXArray {
        let shape = x.shape
        let H = shape[2]
        let W = shape[3]
        let (targetH, targetW) = targetShape

        if H == targetH && W == targetW {
            return x
        }

        // Simple bilinear interpolation
        let yCoords = MLX.linspace(Float(0), Float(H - 1), count: targetH)
        let xCoords = MLX.linspace(Float(0), Float(W - 1), count: targetW)

        let y0 = MLX.floor(yCoords).asType(.int32)
        let y1 = MLX.minimum(y0 + 1, MLXArray(Int32(H - 1)))
        let x0 = MLX.floor(xCoords).asType(.int32)
        let x1 = MLX.minimum(x0 + 1, MLXArray(Int32(W - 1)))

        let wy = (yCoords - y0.asType(.float32)).reshaped([1, 1, targetH, 1])
        let wx = (xCoords - x0.asType(.float32)).reshaped([1, 1, 1, targetW])

        // Sample corners
        let val00 = x[0..., 0..., y0, 0...][0..., 0..., 0..., x0]
        let val01 = x[0..., 0..., y0, 0...][0..., 0..., 0..., x1]
        let val10 = x[0..., 0..., y1, 0...][0..., 0..., 0..., x0]
        let val11 = x[0..., 0..., y1, 0...][0..., 0..., 0..., x1]

        let result = val00 * (1 - wy) * (1 - wx) +
                     val01 * (1 - wy) * wx +
                     val10 * wy * (1 - wx) +
                     val11 * wy * wx

        return result
    }

    /// Simple linear resampling.
    private func resample(_ audio: MLXArray, fromRate: Int, toRate: Int) -> MLXArray {
        if fromRate == toRate {
            return audio
        }

        let batchSize = audio.shape[0]
        let length = audio.shape[1]
        let newLength = length * toRate / fromRate

        var results: [MLXArray] = []
        for b in 0..<batchSize {
            let sample = audio[b]
            let indices = MLX.linspace(Float(0), Float(length - 1), count: newLength)
            let i0 = MLX.floor(indices).asType(.int32)
            let i1 = MLX.minimum(i0 + 1, MLXArray(Int32(length - 1)))
            let w = indices - i0.asType(.float32)

            let v0 = sample[i0]
            let v1 = sample[i1]
            let resampled = v0 * (1 - w) + v1 * w

            results.append(resampled.expandedDimensions(axis: 0))
        }

        return MLX.concatenated(results, axis: 0)
    }
}

File: python/mlx_audio/models/clap/feature_extractor.py
"""CLAP feature extractor for audio preprocessing.

Implements audio preprocessing compatible with HuggingFace's ClapFeatureExtractor.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

import mlx.core as mx
import numpy as np

if TYPE_CHECKING:
    from mlx_audio.models.clap.config import CLAPAudioConfig


class CLAPFeatureExtractor:
    """Audio feature extractor for CLAP models.

    Uses HuggingFace's ClapFeatureExtractor when available for exact
    compatibility. Falls back to a numpy implementation otherwise.

    Args:
        sample_rate: Target sample rate (default: 48000)
        n_fft: FFT window size (default: 1024)
        hop_length: Hop length (default: 480)
        n_mels: Number of mel bins (default: 64)
        max_length_s: Maximum audio length in seconds (default: 10)
    """

    def __init__(
        self,
        sample_rate: int = 48000,
        n_fft: int = 1024,
        hop_length: int = 480,
        n_mels: int = 64,
        max_length_s: float = 10.0,
    ):
        self.sample_rate = sample_rate
        self.n_fft = n_fft
        self.hop_length = hop_length
        self.n_mels = n_mels
        self.max_length_s = max_length_s
        self.nb_max_samples = int(max_length_s * sample_rate)

        # Try to use HuggingFace feature extractor for exact compatibility
        self._hf_processor = None
        try:
            from transformers import ClapProcessor
            self._hf_processor = ClapProcessor.from_pretrained(
                "laion/clap-htsat-fused"
            )
        except (ImportError, Exception):
            pass

    @classmethod
    def from_config(cls, config: "CLAPAudioConfig") -> "CLAPFeatureExtractor":
        """Create feature extractor from CLAP audio config."""
        return cls(
            sample_rate=config.sample_rate,
            n_fft=config.n_fft,
            hop_length=config.hop_length,
            n_mels=config.n_mels,
        )

    def __call__(
        self,
        audio: np.ndarray | mx.array,
        sample_rate: int | None = None,
        return_tensors: str = "mx",
    ) -> dict:
        """Process audio into CLAP-compatible features.

        Args:
            audio: Audio waveform [T] or [B, T]
            sample_rate: Audio sample rate (uses self.sample_rate if None)
            return_tensors: Output format ("mx" or "np")

        Returns:
            Dictionary with:
                - input_features: Mel spectrogram [B, 4, n_frames, n_mels]
                - is_longer: Boolean tensor [B, 1]
        """
        if isinstance(audio, mx.array):
            audio = np.array(audio)

        # Handle batched input
        if audio.ndim == 1:
            audio = audio[np.newaxis, :]

        sr = sample_rate if sample_rate is not None else self.sample_rate

        # Use HuggingFace processor if available (exact compatibility)
        if self._hf_processor is not None:
            return self._process_with_hf(audio, sr, return_tensors)
        else:
            return self._process_fallback(audio, sr, return_tensors)

    def _process_with_hf(
        self,
        audio: np.ndarray,
        sample_rate: int,
        return_tensors: str,
    ) -> dict:
        """Process audio using HuggingFace's feature extractor."""
        batch_size = audio.shape[0]

        # Process each sample
        all_features = []
        all_is_longer = []

        for i in range(batch_size):
            inputs = self._hf_processor(
                audio=audio[i],
                sampling_rate=sample_rate,
                return_tensors="np",
            )
            all_features.append(inputs["input_features"])
            all_is_longer.append(inputs["is_longer"])

        # Stack batch
        input_features = np.concatenate(all_features, axis=0)
        is_longer = np.concatenate(all_is_longer, axis=0)

        if return_tensors == "mx":
            input_features = mx.array(input_features)
            is_longer = mx.array(is_longer)

        return {
            "input_features": input_features,
            "is_longer": is_longer,
        }

    def _process_fallback(
        self,
        audio: np.ndarray,
        sample_rate: int,
        return_tensors: str,
    ) -> dict:
        """Fallback processing without HuggingFace (less accurate)."""
        try:
            import librosa
        except ImportError:
            raise ImportError(
                "librosa is required for audio processing without transformers. "
                "Install with: pip install librosa"
            )

        batch_size = audio.shape[0]
        mel_spectrograms = []
        is_longer_list = []

        for i in range(batch_size):
            sample = audio[i]

            # Resample if needed
            if sample_rate != self.sample_rate:
                sample = librosa.resample(
                    sample,
                    orig_sr=sample_rate,
                    target_sr=self.sample_rate
                )

            # Check if longer than max
            is_longer = len(sample) > self.nb_max_samples
            is_longer_list.append([is_longer])

            # Compute mel spectrogram
            mel = self._compute_mel(sample)

            # For fusion mode, always produce 4 channels
            if is_longer:
                mel = self._compute_fusion_mel(sample)
            else:
                # Pad/truncate and repeat 4 times
                mel = self._compute_padded_mel(sample)
                mel = np.stack([mel, mel, mel, mel], axis=0)

            mel_spectrograms.append(mel)

        input_features = np.stack(mel_spectrograms, axis=0)
        is_longer_arr = np.array(is_longer_list)

        if return_tensors == "mx":
            input_features = mx.array(input_features)
            is_longer_arr = mx.array(is_longer_arr)

        return {
            "input_features": input_features,
            "is_longer": is_longer_arr,
        }

    def _compute_mel(self, audio: np.ndarray) -> np.ndarray:
        """Compute log-mel spectrogram."""
        import librosa

        # Compute mel spectrogram
        mel = librosa.feature.melspectrogram(
            y=audio,
            sr=self.sample_rate,
            n_fft=self.n_fft,
            hop_length=self.hop_length,
            n_mels=self.n_mels,
            fmin=50,
            fmax=14000,
            htk=True,
        )

        # Convert to dB
        mel_db = librosa.power_to_db(mel, ref=1.0, top_db=None)

        return mel_db.T  # [T, F]

    def _compute_padded_mel(self, audio: np.ndarray) -> np.ndarray:
        """Compute mel for audio shorter than max length."""
        # Pad with repeat if needed
        if len(audio) < self.nb_max_samples:
            n_repeat = int(self.nb_max_samples / len(audio))
            audio = np.tile(audio, n_repeat)

        # Pad to exact length
        if len(audio) < self.nb_max_samples:
            audio = np.pad(
                audio,
                (0, self.nb_max_samples - len(audio)),
                mode='constant'
            )
        else:
            audio = audio[:self.nb_max_samples]

        return self._compute_mel(audio)

    def _compute_fusion_mel(self, audio: np.ndarray) -> np.ndarray:
        """Compute fusion mel for long audio (4 channels)."""
        import torch

        # Compute full mel
        mel = self._compute_mel(audio)  # [T, F]

        chunk_frames = self.nb_max_samples // self.hop_length + 1
        total_frames = mel.shape[0]

        if chunk_frames >= total_frames:
            # Audio is actually short enough, just repeat
            return np.stack([mel, mel, mel, mel], axis=0)

        # Get 3 random crops
        ranges = np.array_split(
            list(range(0, total_frames - chunk_frames + 1)), 3
        )
        if len(ranges[1]) == 0:
            ranges[1] = [0]
        if len(ranges[2]) == 0:
            ranges[2] = [0]

        idx_front = np.random.choice(ranges[0])
        idx_middle = np.random.choice(ranges[1])
        idx_back = np.random.choice(ranges[2])

        mel_front = mel[idx_front:idx_front + chunk_frames, :]
        mel_middle = mel[idx_middle:idx_middle + chunk_frames, :]
        mel_back = mel[idx_back:idx_back + chunk_frames, :]

        # Downsampled version
        mel_tensor = torch.tensor(mel[None, None, :])
        mel_shrink = torch.nn.functional.interpolate(
            mel_tensor,
            size=[chunk_frames, self.n_mels],
            mode="bilinear",
            align_corners=False
        )
        mel_shrink = mel_shrink[0][0].numpy()

        return np.stack(
            [mel_shrink, mel_front, mel_middle, mel_back],
            axis=0
        )
